File: 198. House Robber/solution.go
func rob(nums []int) int {
    l := len(nums)
    profit := make([]int, l, l)
    if l == 0 {
        return 0
    } else if l == 1 {
        return nums[0]
    } else if l == 2 {
        return max(nums[0], nums[1]) 
    } 
    
    profit[0] = nums[0]
    profit[1] = nums[1]
    profit[2] = nums[0] + nums[2]
    
    if l == 3 {
        return max(profit[2], profit[1])
    }
    
    for i := 3; i < l; i++ {
        profit[i] = max(profit[i-2], profit[i-3]) + nums[i]
    }
        
    return max(profit[l-2], profit[l-1])
}

func max(a int, b int) int {
    if a > b {
        return a
    } else {
        return b
    }
}

File: 198. House Robber/solution.py
class Solution:
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # edge cases
        if not nums:
            return 0
        if len(nums) < 3:
            return max(nums)
        
        # loot will be calculated bottom up using dynamic programming
        loot = [0]*len(nums)
        for i in range(len(nums)):
            if i - 2 >= 0:
                loot1 = loot[i-2]
            else:
                loot1 = 0
            
            if i - 3 >= 0:
                loot2 = loot[i-3]
            else:
                loot2 = 0                
                
            loot[i] = max(nums[i] + loot1, nums[i] + loot2)
            
        return max(loot[-1], loot[-2])
        
